Move and undo move front-end work

mushtaqak opened the pull request:
# Description

As a part of [Move Epic](https://openedx.atlassian.net/browse/TNL-5943), this PR adds the frontend work to move and undo move an xblock along with Enable/Disable "Move" button for valid/invalid move.

### [TNL-6062](https://openedx.atlassian.net/browse/TNL-6062) 
### [TNL-6063](https://openedx.atlassian.net/browse/TNL-6063)
### [TNL-6229](https://openedx.atlassian.net/browse/TNL-6229)

### [Sandbox link](https://studio-movexblockcomponent.sandbox.edx.org/container/block-v1:edX+DemoX+Demo_Course+type@vertical+block@vertical_0270f6de40fc)

### Testing
- [x] Jasmine
### Reviewers

If you've been tagged for review, please check your corresponding box once you've given the :+1: or approve review.
- [ ] Code review: @muhammad-ammar 
- [x] Code review: @muzaffaryousaf 
### Post-review
- [ ] Squash commits

mushtaqak commented:
@muzaffaryousaf I have created the sandbox build http://jenkins.edx.org:8080/view/Ansible/job/ansible-provision/10907/

muzaffaryousaf commented:
@mushtaqak I think we should increase the confirmation message duration that we show after successful move as well as Undo.

muzaffaryousaf commented:
`"You can not move problem into sequential." ` Can we tell about `"problem"` either if it's component / unit / sequential / chapter.

muzaffaryousaf commented:
> "You can not move problem into sequential." Can we tell about "problem" either if it's component / unit / sequential / chapter.

This could be optional if we don't show this to end users.

mushtaqak commented:
> "You can not move problem into sequential." Can we tell about "problem" either if it's component / unit / sequential / chapter.

Yes, these messages would not be shown to end user once Enable/Disable move button work is done. Besides problem is always a component, I think.

muzaffaryousaf commented:
> @mushtaqak I think we should increase the confirmation message duration that we show after successful move as well as Undo.

@sstack22 @cptvitamin Can any of you suggest the appropriate timeout for confirmation box from product & a11y perspective ?

muzaffaryousaf commented:
> Besides problem is always a component, I think.
You're right but problem vs Component "problem", the later is more clearer than the first one. Since we won't show these messages to end user, we can leave this as it.

muzaffaryousaf commented on common/static/common/js/components/views/feedback_move.js:
```diff
@@ -0,0 +1,57 @@
+                    XBlockViewUtils.moveXBlock(sourceLocator, parentLocator, targetIndex)
+                    .done(function(response) {
+                        // show XBlock element
+                        $('.studio-xblock-wrapper[data-locator="' + response.move_source_locator + '"]').show();
```
What if we don't get `move_source_locator` in response.

mushtaqak replied:
If we are getting response then it must have `move_source_locator` in response.

muzaffaryousaf commented on common/static/common/js/components/views/feedback_move.js:
```diff
@@ -0,0 +1,57 @@
+/**
```
Since `common/static/common/js/components/views/` only contains reusable components, can we move this file to some other appropriate location and change it's name to something `undo_move_xblock`.

muzaffaryousaf replied:
You can leave this as it and move the showMovedBar to some other location.

muzaffaryousaf commented on cms/static/js/views/utils/xblock_utils.js:
```diff
@@ -92,6 +93,36 @@ define(['jquery', 'underscore', 'gettext', 'common/js/components/utils/view_util
+            return ViewUtils.runOperationShowingMessage(operationText,
+                function() {
+                    var movingOperation = $.Deferred();
+                    $.postJSON(ModuleUtils.getUpdateUrl(), {
```
I would call new `$.patchJSON` method instead.

muzaffaryousaf commented on cms/static/js/views/modals/move_xblock_modal.js:
```diff
@@ -88,6 +98,67 @@ function($, Backbone, _, gettext, BaseView, BaseModal, XBlockInfoModel, MoveXBlo
             );
+        },
+
+        showMovedBar: function(title, titleLink, messageLink) {
```
I would name it `showMovedAlert` or `showMovedNotification`.

muzaffaryousaf replied:
I would add doc strings to this method.

muzaffaryousaf commented on cms/static/js/views/modals/move_xblock_modal.js:
```diff
@@ -88,6 +98,67 @@ function($, Backbone, _, gettext, BaseView, BaseModal, XBlockInfoModel, MoveXBlo
+                offset: 0,
+                easing: 'swing',
+                speed: 1000
+            });
```
I'll move L105-120 to a util method to use it in **move** & **undo move**.

muzaffaryousaf commented on cms/static/js/views/modals/move_xblock_modal.js:
```diff
@@ -88,6 +98,67 @@ function($, Backbone, _, gettext, BaseView, BaseModal, XBlockInfoModel, MoveXBlo
+
+        moveXBlock: function() {
+            var self = this;
+            XBlockViewUtils.moveXBlock(self.sourceXBlockInfo.id, self.moveXBlockListView.parent_info.parent.id)
```
Not sure if it's correct way to pick up parent id `self.moveXBlockListView.parent_info.parent.id`

mushtaqak replied:
This would be handled better in `Enable/Disable Move button PR`

muzaffaryousaf commented on cms/static/js/views/modals/move_xblock_modal.js:
```diff
@@ -88,6 +98,67 @@ function($, Backbone, _, gettext, BaseView, BaseModal, XBlockInfoModel, MoveXBlo
+                                {
+                                    displayName: self.sourceXBlockInfo.get('display_name'),
+                                    sourceLocator: self.sourceXBlockInfo.id,
+                                    parentLocator: self.sourceParentXBlockInfo.id,
```
Which parent is this ? source or target ?

muzaffaryousaf replied:
I would rename it to `sourceParentLocator`.

muzaffaryousaf commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -44,9 +43,209 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+
+            sendMoveXBlockRequest = function(requests, xblockLocator, parentLocator, targetIndex, sourceIndex) {
+                var responseData,
+                    expectData,
```
I'll rename it to `expectedData`

muzaffaryousaf commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -44,9 +43,209 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                parent_locator: sourceParentLocator,
+                target_index: sourceIndex
+            });
+            ViewHelpers.verifyNotificationHidden(notificationSpy);
```
I'll check the scroll position / focus too.

muzaffaryousaf commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -44,9 +43,209 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                // select a target item and click on move
+                selectTargetParent(targetParentLocator);
+                modal.$el.find('.modal-actions .action-move').click();
+                verifyNotificationStatus(requests, notificationSpy, 'Moving');
```
I'll also check that notification modal is disappeared after specified value.

mushtaqak replied:
`verifyNotificationStatus` checks that :)

muzaffaryousaf commented:
@mushtaqak I'm done with first pass, let me know when you address them.

muzaffaryousaf commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -44,9 +43,209 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                    target: $(modal.movedAlertView.options.messageLink.text)
+                });
+                verifyNotificationStatus(requests, notificationSpy, 'Undo moving');
+            });
```
@mushtaqak can we confirm the developer error messages via test cases e.g invalid move. I know they're not user facing but i think it's worth to write 1 or 2 test(s) for that.

catong commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -44,9 +43,209 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+
+        getConfirmationFeedbackTitleLink = function(parentLocator) {
+            return StringUtils.interpolate(
+                gettext(' {link_start}Take me there{link_end}'),
```
@mushtaqak Please see the message and link text in https://openedx.atlassian.net/wiki/display/A11Y/Move+UI?preview=/149192768/156542759/Move-13.png

muzaffaryousaf replied:
@catong Are those images final from user facing strings perspective ?

catong replied:
@muzaffaryousaf In this case, probably yes, :-) although there might be other strings I have not seen in context that might cause tweaks to be required. We wanted the link to read more completely "Take me to the new location" rather than "there".

catong commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -44,9 +43,209 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+
+        getUndoConfirmationFeedbackTitle = function(displayName) {
+            return StringUtils.interpolate(
+                gettext('Undo Success! "{sourceDisplayName}" has been moved back to a previous location.'),
```
@mushtaqak Can the success message for "undo move" simply read "Move cancelled. ```{sourceDisplayName}``` has been moved back to its original location."

@chris-mike @sstack22 Your thoughts?

mushtaqak commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -44,9 +43,209 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                        id: parentLocator
+                    }
+                },
+                remove: function() {}   // attach a fake remove method
```
This was added because we needed to test the move functionality on sandbox.

We added this method because we are mocking `moveXBlockListView`, I didn't want to make call to the actual `moveXBlockListView` which would require extra arguments and would more complex. 

But, ultimately this needs to be changed in Enable/Disable Move ( we would get `targetParentInfo` in a better way and in this test change this method accordingly ...

cptvitamin commented:
> @sstack22 @cptvitamin Can any of you suggest the appropriate timeout for confirmation box from product & a11y perspective ?

@muzaffaryousaf I'm not familiar with the type of notification this is.  Is it the one in the mockup/screenshot in https://openedx.atlassian.net/browse/TNL-6229 ? That one has an "X" to close or dismiss the notification. Is it possible to require the user to dismiss the notification in this case as well?

See https://www.w3.org/TR/wai-aria-practices/#alert and https://www.w3.org/TR/UNDERSTANDING-WCAG20/time-limits-no-exceptions.html

Both recommend not automatically timing out.

mushtaqak commented:
@muzaffaryousaf Sandbox is updated :)

muzaffaryousaf commented:
> @muzaffaryousaf I'm not familiar with the type of notification this is. Is it the one in the mockup/screenshot in https://openedx.atlassian.net/browse/TNL-6229 ? That one has an "X" to close or dismiss the notification. Is it possible to require the user to dismiss the notification in this case as well?
See https://www.w3.org/TR/wai-aria-practices/#alert and https://www.w3.org/TR/UNDERSTANDING-WCAG20/time-limits-no-exceptions.html
Both recommend not automatically timing out.

@cptvitamin It's similar with auto timeout just like Advance Settings page that we were suppose to follow. 
@sstack22 Thoughts ???

muzaffaryousaf commented:
> @cptvitamin It's similar with auto timeout just like Advance Settings page that we were suppose to follow.
@sstack22 Thoughts ???

@cptvitamin @sstack22 According to the a11y docs, i think we should not use auto timeout for notification.

sstack22 commented:
@muzaffaryousaf - is that for the 'moving' message? how long do we expect it to take a component to move?

muzaffaryousaf commented:
@sstack22 No, it's for success message where we show undo and take me there link.

sstack22 commented:
@muzaffaryousaf - let's not time out the message. The 'X' looks a bit odd though. CC: @chris-mike

sstack22 commented:
@muzaffaryousaf - apologies for the churn, but I think the back button is redundant ( which @chris-mike did previously point out to me before, but I requested :) ). Let's remove the back button, which I think enables us remove the padding around modal content.

@chris-mike , @marcotuts - re: breadcrumbs - I recall we had a similar, nested pattern in the discussion forum at some point, but it looks like we removed it. Any context on the change? I honestly would prefer the breadcrumbs to be on the same line, since it'll free up more space and I think the clarity of where you are is there (via clickable breadcrumbs, and the section/subsection info). My preference would be to start with 1 line, and then move to nested if needed based on feedback.

chris-mike commented:
Looking good.  I have a few design changes to suggest.

1. Minimize whitespace between grey modal frame and move section. Padding for top, left and right need not be more than 10px. Padding on bottom is fine (see screenshot)
2. In the breadcrumbs/upper section, unlinked text that reads "Course Outline" or whatever name of the current section should be 18px (this is the black, unlinked text). So, upon initial view, "course outline" should be bigger.  This sizing should be consistent as you progress down the chain.  (The linked content is the right size.)
3. Re: the "back" arrow, can you make it look more disabled at Course outline level: #d9d9d9 (palette(grayscale, back))
4. Can we change background color of the breadcrumbs/upper section  to  palette(primary, x-back) RGB: rgb(242, 248, 251) HEX: f2f8fb (see screenshot). There wasn't enough to help the eye focus.
5. In the section/subsection/units/components area, remove indentation and left-align all of the text. So, "units" label needs to move right about 10px, to line up with the unit names.  all in line with the back arrow.
6. In the section/subsection/units area, I'd prefer to see the text and arrows a different color: palette(primary, dark) RGB: rgb(6, 86, 131) HEX: 065683.  This will still have contrast when hovered.  
7. The disabled "move" button - is this color the standard for Studio?  I've not seen it before and it seems to be off-brand. I'd prefer a disabled grey.
8. Cancel should underline on hover.  This blue also looks to be the wrong color, but if a Studio default, then it's OK.

The hover looks great. 

![screen shot 2017-01-26 at 4 41 17 pm](https://cloud.githubusercontent.com/assets/16578372/22351508/4d34004e-e3e6-11e6-949b-4b4573b61310.png)

Regarding the modal itself:
1. When you open the modal, it seems to be lower than I expected. @sstack22 What is standard for Studio?  I'd expect it to be closer to where "edit" appears, which seems to be 30% from top of page (Move seems to be 60% from top of page).  I'd like to see it open higher on the page, at the same level of edit.
2. Can you add the stacked FA move icon (folder/arrow) to the component?

chris-mike commented:
One more. When tabbing through the onfocus loses the blue background color seen on hover.
![screen shot 2017-01-26 at 4 52 39 pm](https://cloud.githubusercontent.com/assets/16578372/22351897/eb4fa8f4-e3e7-11e6-992c-e77e6a9c31e0.png)

chris-mike commented:
OK, one more... this one for @sstack22 .  Can we tighten line height a bit of the nested breadcrumbs?  Particularly the ones that are linked.

marcotuts commented:
@chris-mike  - I would echo @sstack22's feedback here on the nested breadcrumbs model. I would suggest we merge a horizontal breadcrumb pattern that uses 1 font awesome icon (currently we have angle, caret, and arrow). I'm concerned about the usability and visual height afforded to what should be a secondary detail provided you are making the correct decisions as you navigate into the course tree.  Sandbox browser hacked below for context:
![image](https://cloud.githubusercontent.com/assets/2023680/22352734/580c66a0-e3eb-11e6-99a3-57841b8d2c2d.png)

mushtaqak commented:
Sandbox has been updated with latest changes :)

@chris-mike All of your feedback is addressed except Cancel button underline which does not appear on any other modals in studio (example: see Edit modal on unit page)

@sstack22 @chris-mike @marcotuts We are also waiting on your final decision on breadcrumb design, Back arrow button and Move/Undo confirmation message close icon.

FYI @muzaffaryousaf @muhammad-ammar

chris-mike commented:
1. As you go down the tree, the bottom most branch should be 18px (same size of "course outline" at top of branch).
![movebottom-branch](https://cloud.githubusercontent.com/assets/16578372/22388986/5fad65fa-e4b1-11e6-987c-e036d92918e6.png)

2. The current location indicator should be smaller and black. Bring it down a few pts. in size.

3. Occasionally, there's extra spacing between the componets label and the units list. See screen. Not sure why.
![screen shot 2017-01-27 at 4 59 04 pm](https://cloud.githubusercontent.com/assets/16578372/22389107/01125a7c-e4b2-11e6-98fe-538502345e69.png)

4. when you show components, they should not be blue, they should be grey. We need the color diff. to enforce the idea that they are not move options. Can't get to confluence to get exact hex. I think it was:  palette(grayscale, base) RGB: rgb(65, 65, 65) HEX: 414141

5. Can you remove the bottom border from the breadcrumb-container. No longer needed with blue background.

@marcotuts @sstack22 With regards to breadcrumb nesting, I prefer current design.  I think to put everything on one line means we will wrap text and it will look like one long run on sentence.

chris-mike commented:
Move isn't working so I cannot see the success messaging.

sstack22 commented:
@chris-mike - just did a bit of research today on typical title length (yay for course graph and adam + lily!). For all levels, the majority of titles are 10-25 characters (including whitespace).  I don't think the 1 line would cause an issue with wrapping, and we could truncate the titles with ellipsis to be sure of that (especially to account for the upperbounds of title length). Given the avg. title length, I do think we should keep to the standard pattern. 

also @muzaffaryousaf , @chris-mike - was the move icon changed purposefully? i might be missing a comment on it.

muzaffaryousaf commented:
> Move isn't working so I cannot see the success messaging.

@chris-mike I'm able to move component and can also see the success message, not sure if you're facing any error ?

muzaffaryousaf commented:
> also @muzaffaryousaf , @chris-mike - was the move icon changed purposefully? i might be missing a comment on it.

@sstack22 That was an request from @chris-mike in Emerald group.

muhammad-ammar commented:
> Occasionally, there's extra spacing between the componets label and the units list. See screen. Not sure why.

@chris-mike 
The extra space is because [first component](
https://studio-movexblockcomponent.sandbox.edx.org/container/block-v1:edX+DemoX+Demo_Course+type@vertical+block@4f6c1b4e316a419ab5b6bf30e6c708e9) on that unit page has not display name.

muhammad-ammar commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -53,11 +52,231 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                var responseData,
+                    expectedData,
+                    sourceIndex = sourceIndex || 0, // eslint-disable-line no-redeclare
+                    moveButton = modal.$el.find('.modal-actions .action-move')[sourceIndex];
```
I think we don't need `sourceIndex`

muhammad-ammar commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -53,11 +52,231 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                moveXBlockWithSuccess(requests);
+            });
+
+            it('undo move an xblock when undo move button is clicked', function() {
```
`undo move link is clicked`

muhammad-ammar commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -53,11 +52,231 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                moveXBlockWithSuccess(requests);
+                modal.movedAlertView.undoMoveXBlock({
+                    target: $(modal.movedAlertView.options.messageHtml.text)
+                });
```
We should click on undo move link instead of directly calling the `undoMoveXBlock` method.

muhammad-ammar commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -53,11 +52,231 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                // select a target parent and click cancel button
+                selectTargetParent(targetParentLocator);
+                modal.$el.find('.modal-actions .action-cancel')[sourceIndex].click();
+                expect(modal.movedAlertView).toBeNull();
```
We also need to verify that no request is sent.

muhammad-ammar commented on cms/static/js/spec/views/modals/move_xblock_modal_spec.js:
```diff
@@ -53,11 +52,231 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                    parent_locator: sourceParentLocator,
+                    target_index: sourceIndex
+                });
+                expect(modal.movedAlertView.movedAlertView.options.title).toEqual(
```
We should verify the rendered message instead of using `modal.movedAlertView.movedAlertView.options.title`

mushtaqak commented:
@chris-mike I have started to update the sandbox, it would probably take 50 minutes from now. 

@muhammad-ammar has addressed your feedback, kindly have a look. Thanks :)

sstack22 commented:
@muzaffaryousaf , @chris-mike - given the data we found re: title lengths, let's go with 1 line for now (back-button not needed), and add truncation in a separate PR to account for the edge cases and ensure no wrapping. Let's revisit nesting breadcrumbs in the future as necessary. 

Also, thanks for clarifying the icon. Very cool that you can nest FA icons!

mushtaqak commented on cms/static/js/views/modals/move_xblock_modal.js:
```diff
@@ -28,20 +37,26 @@ function($, Backbone, _, gettext, BaseView, BaseModal, XBlockInfoModel, MoveXBlo
             this.options.title = this.getTitle();
+            this.movedAlertView = null;
+            this.moveXBlockBreadcrumbView = null;
+            this.moveXBlockListView = null;
```
We are setting null because we will be using these views across the move flow.

sstack22 commented:
@chris-mike , @muzaffaryousaf - re: confirmation message. Should the 'undo' and 'take me there' be buttons? The message looks like it's missing a bit of styling.

muzaffaryousaf commented:
@sstack22 I'm not sure if you're seeing this message properly 
<img width="1280" alt="screen shot 2017-01-31 at 7 22 57 pm" src="https://cloud.githubusercontent.com/assets/2994509/22468616/c6177498-e7ea-11e6-8b94-8e881562d02b.png">

mushtaqak commented:
@muzaffaryousaf @muhammad-ammar addressed feedbacks and added new tests :)

@all sandbox is updated :)

muzaffaryousaf commented:
@mushtaqak Make sure you do rebase and update sandbox at day end.

mushtaqak commented:
@cptvitamin Thanks for your input. 

> Can we move focus to the confirmation message container

Yes, I will do that

> make the container an aria-live region? 
> Can we make the Close button an actual button and not a link?

We are using existing `SystemFeedback` alert which has close link, so changing this might affect other places as well and that might end up out of the scope of this PR. 

>Can we fix the contrast of the links in the confirmation message? Currently, blue on black is not accessible.

Yes, @chris-mike provided the colors, I will apply them.

> Can we move this on the same line as "Undo move"? Currently its included inside the< h2 > element and we don't typically include links inside Headings.

Let's see what we decide @muzaffaryousaf @sstack22 FYI

muhammad-ammar commented:
@cptvitamin Which element should be focused after we navigate down or up?

cptvitamin commented:
@muhammad-ammar Can you move focus to the dialog container itself?  The element that receives focus when the dialog box opens.

mushtaqak commented:
@chris-mike @cptvitamin @sstack22 I did some tweaks to browser, Here is how banner would look with different scenarios :

<img width="809" alt="success-message-banner" src="https://cloud.githubusercontent.com/assets/6991154/22686914/e0cce478-ed47-11e6-91f4-89ca72c86df8.png">

This is how success banner would look like with provided colors.
P.S: `Take me to new location` is in hover state, `Undo move` is in normal state. I would suggest to have blue color (previously implemented) to be in normal state.

<img width="1213" alt="screen shot 2017-02-07 at 3 25 40 pm" src="https://cloud.githubusercontent.com/assets/6991154/22687432/d21544a0-ed49-11e6-8ae3-8de9589fcd0b.png">
This is how banner would look with adding action buttons.

P.S: @cptvitamin `Take me to new location` actually redirects us to a new location, we would have to redirect now from front-end. Is it accessible if we have button but we work as link(redirect functionality)


FYI:  @muzaffaryousaf @muhammad-ammar

muzaffaryousaf commented:
> screen shot 2017-02-07 at 3 25 40 pm This is how banner would look with adding action buttons.
P.S: @cptvitamin Take me to new location actually redirects us to a new location, we would have to redirect now from front-end. Is it accessible if we have button but we work as link(redirect functionality)

@sstack22 @cptvitamin Looks like the existing banner that we use on Advance Settings page won't be accessible in this case since banner uses Primary/Secondary buttons and for move we need 1 link for "Take me to the new location". I'd say that first option makes more sense i.e. moving the actions on separate line with given contrast. Let me know what do you think. Thanks.

cptvitamin commented:
@mushtaqak @muzaffaryousaf I wouldn't get hung up on whether its a button or a link in this case since *technically* Undo is an action, and Take me there is a link.  What's important is that both are the same.  If we go with solution number one, we may want to convert the two links to an unordered list, with visible bullet points.  This makes the two available actions stand out a little more from the Heading since all the text is relatively the same size.  This is looking good though.  Thanks for your work on this.

muzaffaryousaf commented:
@sstack22 Please share your thoughts too.

sstack22 commented:
@mushtaqak @muzaffaryousaf @cptvitamin, let's go with option 2 (the button). Note that undo should be a button, and take me there a link. Let's also confirm the color contrast of the link text meets a11y requirements.

Thanks for all of your work on this! It looks awesome. :)

muzaffaryousaf commented:
@cptvitamin Do you guys use any tool for color contrast selection ?

mushtaqak commented:
> @mushtaqak @muzaffaryousaf @cptvitamin, let's go with option 2 (the button). Note that undo should be a button, and take me there a link. Let's also confirm the color contrast of the link text meets a11y requirements.

So `Undo move` would be a button and `take me to the new location` a link. So `Undo move` would come before the `take me to the new location` link, right @sstack22 ?

mushtaqak commented:
@cptvitamin @muhammad-ammar has addressed your feedback :)

Sandbox updated. FYI @sstack22 @chris-mike @muzaffaryousaf

chris-mike commented:
@muzaffaryousaf I use this site for contrast checking: http://webaim.org/resources/contrastchecker/.

I like the button/link approach. We'd definitely need to change the green background on this (current color is not WCAG 2.0).  This is the right green for background with white text:
 palette(success, text) RGB: rgb(0, 129, 0) HEX: 008100

chris-mike commented:
I see no other issues at this time. Thx.

mushtaqak commented:
> palette(success, text) RGB: rgb(0, 129, 0) HEX: 008100

@chris-mike we are already using this exact background color for the button.

mushtaqak commented:
>Move and cancel buttons
Can we add SR text "here" to the button?
Can we make both the Move and Cancel buttons actual buttons instead of links?
Once these are buttons, can we toggle a disabled attribute on the Move button appropriately? Currently, there is no indication to a non-visual user that the button is disabled. Keyboard users can still click the button too.

@cptvitamin We are using an existing modal, I think this would be out of the scope of this PR but we can do this in a separate ticket may be :)

mushtaqak commented:
@sstack22 @cptvitamin Sandbox has been updated, please have a look :)

FYI: @muzaffaryousaf  @muhammad-ammar

muzaffaryousaf commented:
jenkins run bokchoy

muhammad-ammar commented:
jenkins run bokchoy

muhammad-ammar commented on cms/static/js/spec/views/move_xblock_spec.js:
```diff
@@ -336,9 +367,226 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                    expect(modal.movedAlertView).toBeDefined();
+                    verifyConfirmationFeedbackTitle(sourceDisplayName);
+                    verifyConfirmationFeedbackRedirectLink(sourceDisplayName);
+                    verifyConfirmationFeedbackUndoMoveAction(
```
Can we rename above three methods and add Html in the end because all these methods are verifying rendered html.

`verifyConfirmationFeedbackTitle` -> `verifyConfirmationFeedbackTitleHtml`
`verifyConfirmationFeedbackRedirectLink` -> `verifyConfirmationFeedbackRedirectLinkHtml`
`verifyConfirmationFeedbackUndoMoveAction` -> `verifyConfirmationFeedbackUndoMoveActionHtml`

muzaffaryousaf commented:
@sstack22 @cptvitamin We've addressed all of the feedback that you've provided. Can you please take another look and let us know if we need to address anything else. Thanks.

muzaffaryousaf commented on cms/static/js/views/modals/move_xblock_modal.js:
```diff
@@ -6,14 +6,21 @@ define([
     'js/models/xblock_info', 'js/views/move_xblock_list', 'js/views/move_xblock_breadcrumb',
     'common/js/components/views/feedback',
+    'js/views/utils/xblock_utils',
+    'js/views/utils/move_xblock_utils',
```
I'd add feature based items after the util / helper methods.

muzaffaryousaf commented on cms/static/js/spec/views/move_xblock_spec.js:
```diff
@@ -336,9 +367,226 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                    moveXBlockWithSuccess(requests);
+                    notificationSpy = ViewHelpers.createNotificationSpy();
+                    modal.movedAlertView.$el.find('.action-save').click();
+                    verifyNotificationStatus(requests, notificationSpy, 'Undo moving');
```
@mushtaqak I'd like to see some tests for failure case. e.g. `Undo Move returned 500`, ` take me there returned 500`, `Actual move returned 500`.

muzaffaryousaf commented on cms/static/js/views/utils/move_xblock_utils.js:
```diff
@@ -0,0 +1,88 @@
+            return movedAlertView;
+        };
+
+        hideMovedNotification = function(SystemFeedback) {
```
I think we should't have `SystemFeedback` as an argument, can't we define Feedback in this file and use that in this method instead of parameter?

muzaffaryousaf commented on cms/static/js/views/utils/move_xblock_utils.js:
```diff
@@ -0,0 +1,88 @@
+
+        showMovedNotification = function(title, data) {
+            var movedAlertView;
+            if (data) {
```
I did't get why we need this check and else part ?

mushtaqak replied:
it's because for undo move cancelled notification, we are not proving any extra data.

muzaffaryousaf commented on cms/static/js/views/utils/xblock_utils.js:
```diff
@@ -92,6 +93,34 @@ define(['jquery', 'underscore', 'gettext', 'common/js/components/utils/view_util
 
         /**
+         * Moves the specified xblock in a new parent xblock.
+         * @param {String}  sourceLocator  The xblock element to be moved.
```
`xblock element` --> `Locator of xblock element`.

muzaffaryousaf commented on cms/static/sass/elements/_modal-window.scss:
```diff
@@ -297,6 +297,11 @@
+    .modal-actions .action-move.is-disabled {
+      border: 1px solid $gray-l1 !important;
+      background: $gray-l1 !important;
+    }
```
@mushtaqak Don't we have existing disabled styles in scss ? I believe we're already using disabled buttons in studio?

mushtaqak replied:
Yes, but modal disabled was a bit different than the normal studio button disabled, so on review I had to change this.

mushtaqak commented on cms/static/js/spec/views/move_xblock_spec.js:
```diff
@@ -24,21 +27,61 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
                 TemplateHelpers.installTemplates([
-                    'move-xblock-list',
-                    'move-xblock-breadcrumb'
+                    'basic-modal',
```
we needed this because not all things were loaded correctly, also this was used previously in `move_xblock_modal_spec.js`

mushtaqak commented on cms/static/js/spec/views/move_xblock_spec.js:
```diff
@@ -336,9 +367,226 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                it('moves an xblock when move button is clicked', function() {
+                    var requests = AjaxHelpers.requests(this);
+                    moveXBlockWithSuccess(requests);
+                });
```
I think we need to test if an item is moved on move button click.

muzaffaryousaf replied:
We're already doing the same stuff in other tests.

mushtaqak commented:
> Once these are buttons, can we toggle a disabled attribute on the Move button appropriately? Currently, there is no indication to a non-visual user that the button is disabled. Keyboard users can still click the button too.

@cptvitamin We are using existing modal which is used across the studio, these buttons are actually anchor tags, so currently I can't add disabled attribute unless they are buttons which is out of the scope of this PR.

FYI: @muzaffaryousaf

mushtaqak commented:
@muhammad-ammar @muzaffaryousaf feedback addressed :)

mushtaqak commented:
@sstack22 @cptvitamin Please have look :)

sstack22 commented:
This is awesome! 👍 on my end.

cptvitamin commented:
@mushtaqak The issues I pointed out with the move and cancel button must be fixed at some point, so lets capture those in another ticket.

cptvitamin commented on cms/static/js/views/modals/move_xblock_modal.js:
```diff
@@ -105,6 +125,53 @@ function($, Backbone, _, gettext, BaseView, BaseModal, XBlockInfoModel, MoveXBlo
+        updateMoveState: function(isValidMove) {
+            var $moveButton = this.$el.find('.action-move');
+            if (isValidMove) {
+                $moveButton.removeAttr('disabled').removeClass('is-disabled');
```
since this is not actually a button, it cannot have a disabled attribute.  It is currently not doing anything.  This will need to be a proper button for this to work.

muzaffaryousaf commented:
> @mushtaqak The issues I pointed out with the move and cancel button must be fixed at some point, so lets capture those in another ticket.

@mushtaqak Let's create a ticket for save and cancel button of the dialogue box. Also, mention about disabled attribute.

muzaffaryousaf commented on cms/static/js/views/utils/move_xblock_utils.js:
```diff
@@ -0,0 +1,106 @@
+            speed: 1000
+        });
+        movedAlertView.$('.wrapper').first().focus();
+        return movedAlertView;
```
Do we really need to return alert ?

muzaffaryousaf replied:
I see you're using it in move method.

muzaffaryousaf commented on cms/static/js/spec/views/move_xblock_spec.js:
```diff
@@ -580,14 +553,36 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
                 });
 
-                it('shows a notification when undo moving', function() {
+                it('show a notification when undo moving', function() {
```
I'd recommend use proper naming convention for tests. https://jasmine.github.io/2.0/introduction.html#section-Grouping_Related_Specs_with_<code>describe</code>

muzaffaryousaf commented:
@mushtaqak I'm done with 2nd pass, things are in good shape now, Thanks for great work. 👍

muzaffaryousaf commented:
@mushtaqak Don't forget to squash the commits.

muzaffaryousaf commented on cms/static/js/spec/views/move_xblock_spec.js:
```diff
@@ -336,9 +367,226 @@ define(['jquery', 'underscore', 'edx-ui-toolkit/js/utils/spec-helpers/ajax-helpe
+                });
+            });
+
+            describe('Move an xblock', function() {
```
`'Move an xblock'` still does't make any sense better to rename it.

muzaffaryousaf replied:
you can refactor in 2nd PR.

muzaffaryousaf commented:
jenkins run python

muzaffaryousaf commented:
LGTM, squash, get green and :shipit:

mushtaqak commented:
jenkins run python
